monix-eval/shared/src/main/scala/monix/eval/Task.scala

```diff
@@ -783,54 +801,57 @@ object Task extends TaskInstances {
     *
     * Although the effects are unordered, we ensure the order of results
     * matches the order of the input sequence. Also see [[gatherUnordered]].
-    *
-    * Alias for [[zipList]].
     */
-  def gather[A](in: Seq[Task[A]]): Task[List[A]] =
-    zipList(in)
+  def gather[A, M[X] <: TraversableOnce[X]](in: M[Task[A]])
+    (implicit cbf: CanBuildFrom[M[Task[A]], A, M[A]]): Task[M[A]] = {
+    val init = evalAlways(cbf(in))
+    val r = in.foldLeft(init)((acc,elem) => Task.mapBoth(acc,elem)(_ += _))
+    r.map(_.result())
+  }
 
-  /** Nondeterministically gather results from the given sequence of tasks
-    * to a sequence, without keeping the original ordering of results.
+  /** Nondeterministically gather results from the given collection of tasks,
+    * without keeping the original ordering of results.
     *
     * This function is similar to [[gather]], but neither the effects nor the
     * results will be ordered. Useful when you don't need ordering because it
     * can be more efficient than `gather`.
     */
-  def gatherUnordered[A](in: Seq[Task[A]]): Task[Seq[A]] =
+  def gatherUnordered[A, M[X] <: TraversableOnce[X]](in: M[Task[A]])
+    (implicit cbf: CanBuildFrom[M[Task[A]], A, M[A]]): Task[M[A]] =
     Async { (scheduler, conn, finalCallback) =>
-      val atom = Atomic(Vector.empty[A])
-      val expected = in.length
+      var remaining = in.size
```

Comment:
Don't worry, we don't usually work with TraversableOnce, that's why :-)

Doing `var remaining = 0` is not sufficient because `lock.synchronize` is reentrant and some tasks can execute immediately on the same thread (e.g. `evalAlways`, `evalOnce`, `now`). And we are starting these tasks with `unsafeStartNow`, so serial execution is possible. If we would have started them with `unsafeStartAsync`, then your point would be correct.